scripts/device_check.sh

```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
+
+# This script is designed to detect the number of gpu devices that are available before the
+# ctest suite is run. It does this by using the nvidia-smi. The script is only called if
+# parthenon is built with kokkos_ENABLE_CUDA.
+#
+# The script takes a single argument the number of GPUs that are meant to be used with the 
+# tests. The script then checks that the number of GPUs that are actually avaliable on the
+# system are enough to satisfy this requirement. 
+
+if [ "$#" -ne 1 ]; then
+  printf "You must enter exactly 1 command line arguments, which should indicate
+the number of GPUs to be used with the tests.\n"
+  exit 1
+fi
+
+NUM_GPUS_ASSIGNED_FOR_TEST=$1
+
+if ! command -v nvidia-smi &> /dev/null
+then
+  printf "CUDA has been enabled but nvidia-smi cannot be found\n"
+  exit 1 
+fi
+
+output=$(nvidia-smi -L)
+
+found=$(echo $output | grep "GPU 0: ")
+
+GPUS_DETECTED=0
+while [ ! -z "$found" ]; do
+  let GPUS_DETECTED+=1
+  found=$(echo $output | grep "GPU ${GPUS_DETECTED}: ")
+done
+
+if [ "$GPUS_DETECTED" -eq "0" ]
+then
+  printf "CUDA has been enabled but no GPUs have been detected.\n"
+  exit 1
+elif [ "$NUM_GPUS_ASSIGNED_FOR_TEST" -gt "$GPUS_DETECTED" ]
+then
+  printf "You are trying to build the parthenon regression 
+tests with CUDA enabled kokkos, with the following settings:\n
+Number of CUDA devices set to: ${NUM_GPUS_ASSIGNED_FOR_TEST}
+Number of CUDA devices available: ${GPUS_DETECTED}\n
+The number of gpus detected is less than then the number of devices requested,
+consider changing:\n
+NUM_GPU_DEVICES_TESTING=${GPUS_DETECTED}
+NUM_MPI_PROC_TESTING=${GPUS_DETECTED}\n
+Or consider building without CUDA.\n"
+  exit 1
+elif [ "$NUM_GPUS_ASSIGNED_FOR_TEST" -lt "$GPUS_DETECTED" ]
+then
+  printf "\nNumber of GPUs detected: $GPUS_DETECTED
```

Comment:
Hm... this warning seems kind of annoying and unnecessary - I would expect running tests with 1 MPI rank on a Sierra or Summit-like node to be common. Either get rid of this only output this warning if `NUM_GPUS_ASSIGNED_FOR_TEST` does not equal `NUM_MPI_PROC_TESTING`. If this has already been hashed out, feel free to ignore.

Reply:
This is not a warning, and does not appear during the config step. It just helps the user to know what values are actually being used when the tests are being run, this is not triggered by cmake, the script is called from within the ctest framework. I think it is worth keeping as cmake and ctest are not guaranteed to have the same values.

Reply:
Disregard first comment, I only saw the top line that was highlighted.

Reply:
Ok, fair point.